sdk/core/azure-core/inc/http/bodyStream.hpp

```diff
@@ -0,0 +1,269 @@
+
+namespace Azure { namespace Core { namespace Http {
+
+  // BodyStream is used to read data to/from a service
+  class BodyStream {
+  public:
+    virtual ~BodyStream() = 0;
+
+    // Returns the length of the data; used with the HTTP Content-Length header
+    virtual int64_t Length() const = 0;
+
+    // Resets the stream back to the beginning (for retries)
+    // Derived classes that send data in an HTTP request MUST override this and implement it
+    // properly.
+    virtual void Rewind()
+    {
+      throw "Not Implemented"; // TODO: Replace with best practice as defined by guideline
+    };
+
+    // Reads more data; throws if error/canceled
+    // return copied size
+    virtual int64_t Read(Context& context, uint8_t* buffer, int64_t count) = 0;
+
+    // Closes the stream; typically called after all data read or if an error occurs.
+    virtual void Close() = 0;
+
+    // Keep reading until buffer is all fill out of the end of stream content is reached
+    static int64_t ReadToCount(Context& context, BodyStream& body, uint8_t* buffer, int64_t count)
+    {
+      int64_t readBytes;
+      int64_t totalRead = 0;
+
+      for (;;)
+      {
+        readBytes = body.Read(context, buffer + totalRead, count - totalRead);
+        totalRead += readBytes;
+        // Reach all of buffer size
+        if (totalRead == count || readBytes == 0)
+        {
+          return totalRead;
+        }
+      }
+    }
+
+    static std::unique_ptr<std::vector<uint8_t>> ReadToEnd(Context& context, BodyStream& body)
+    {
+      constexpr int64_t chunkSize = 1024 * 8;
+      auto unique_buffer = std::make_unique<std::vector<uint8_t>>();
+
+      for (auto chunkNumber = 0;; chunkNumber++)
+      {
+        unique_buffer->resize((chunkNumber + 1) * chunkSize);
+        int64_t readBytes = ReadToCount(
+            context, body, unique_buffer->data() + (chunkNumber * chunkSize), chunkSize);
+
+        if (readBytes < chunkSize)
+        {
+          unique_buffer->resize((chunkNumber * chunkSize) + readBytes);
+          return unique_buffer;
+        }
+      }
+    }
+  };
+
+  class MemoryBodyStream : public BodyStream {
+  private:
+    const uint8_t* m_data;
+    int64_t m_length;
+    int64_t m_offset = 0;
+
+  public:
+    // Forbid constructor for rval so we don't end up storing dangling ptr
+    MemoryBodyStream(std::vector<uint8_t> const&&) = delete;
+
+    MemoryBodyStream(std::vector<uint8_t> const& buffer)
+        : MemoryBodyStream(buffer.data(), static_cast<uint64_t>(buffer.size()))
+    {
+    }
+
+    // cast as vector from ptr and length
+    explicit MemoryBodyStream(const uint8_t* data, int64_t length) : m_data(data), m_length(length)
+    {
+    }
+
+    int64_t Length() const override { return this->m_length; }
+
+    int64_t Read(Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
```

Comment:
One conclusion from Jun. 10 meeting is

* Use C++ style casts
* Avoid C casts.

reference to the recording at around 52 min.

Reply:
ditto for all other casts, since I've seen these C style casts a lot in your code.